src/Utilities/CssConcatenator/RuleBlock.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pelago\Emogrifier\Utilities\CssConcatenator;
+
+/**
+ * This class represents a CSS rule block.
+ *
+ * @internal
+ *
+ * @author Jake Hotson <jake.github@qzdesign.co.uk>
+ * @author SignpostMarv
+ * @author Oliver Klee <github@oliverklee.de>
+ */
+class RuleBlock
+{
+    /**
+     * array whose keys are selectors for the rule block (values are of no significance)
+     *
+     * @var array
+     *
+     * @psalm-var array<string, mixed>
+     */
+    private $selectorsAsKeys = [];
+
+    /**
+     * the property declarations, e.g. "margin-top: 0.5em; padding: 0"
+     *
+     * @var string
+     */
+    private $declarationsBlock;
+
+    /**
+     * @param array $selectorsAsKeys
+     * @param string $declarationsBlock
+     *
+     * @psalm-param array<string, mixed> $selectorsAsKeys
+     */
+    public function __construct(array $selectorsAsKeys, string $declarationsBlock)
```

Comment:
@JakeQZ Does it make sense to expect the selectors as keys here, or can we move the key/value reversal into this class?

Reply:
tbh, I was a little confused as to why the selectors are retained as keys, as opposed to `array<int, string>`.

Reply:
The reason for them being stored as keys is to be able to more efficiently (i.e. in O(nlogn) rather than O(n2) time) do the following:
- combine a set of selectors without ending up with duplicates;
- test if two sets of selectors are equivalent, but possibly in a different order.

But looking at `CssConcatenator::append()`, it looks like we could move the bulk of the logic there into this new class.  We could have a method named something like `maybeMerge` which is passed another `RuleBlock` object, and if possible combines `$that` with `$this`, returning `true` if it was able to and has done so.  If it returns `false`, `CssConcatenator::append()` would then add it to the end of `$mediaRule->ruleBlocks`.  (If the new `RuleBlock` was merged, it would only exist for the lifetime of `append()`, being temporarily created to be passed as the parameter to `maybeMerge()`.)

Taking this approach would allow the constructor to be passed the selectors as a normal 'unflipped' array of strings, but store them internally in its class property as the result of `array_flip()`.  I think it would tidy the logic up quite nicely...